Source/CarthageKit/Project.swift

```diff
@@ -1165,6 +1183,173 @@ public final class Project { // swiftlint:disable:this type_body_length
+				// Check if the submodule is initialized/updated already.
+				return isRepository && FileManager.default.fileExists(atPath: submoduleDirectoryURL.appendingPathComponent(".git").path)
+			}
+			.flatMap(.merge) { submoduleExists -> SignalProducer<(), CarthageError> in
+				if submoduleExists {
+					// Just check out and stage the correct revision.
+					return fetchRepository(submoduleDirectoryURL, remoteURL: fetchURL, refspec: "+refs/heads/*:refs/remotes/origin/*")
+						.then(
+							launchGitTask(
+								["config", "--file", ".gitmodules", "submodule.\(submodule.name).url", submodule.url.urlString],
+								repositoryFileURL: repositoryFileURL
+							)
+						)
+						.then(launchGitTask([ "submodule", "--quiet", "sync", "--recursive", submoduleDirectoryURL.path ], repositoryFileURL: repositoryFileURL))
+						.then(checkoutRepositoryToDirectory(submoduleDirectoryURL, submoduleDirectoryURL, force: false))
+						.then(launchGitTask([ "add", "--force", submodule.path ], repositoryFileURL: repositoryFileURL))
+						.startOnQueue(RepositoryOperationQueues.queue(forLocalRepositoryURL: submoduleDirectoryURL))
+						.then(self.updateSubmodules(submoduleDirectoryURL))
+				} else {
+					// `git clone` will fail if there's an existing file at that path, so try to remove
+					// anything that's currently there. If this fails, then we're no worse off.
+					// This can happen if you first do `carthage checkout` and then try it again with
+					// `--use-submodules`, e.g.
+					if FileManager.default.fileExists(atPath: submoduleDirectoryURL.path) {
+						try? FileManager.default.removeItem(at: submoduleDirectoryURL)
+					}
+
+					let addSubmodule = launchGitTask(
+							["submodule", "--quiet", "add", "--force", "--name", submodule.name, "--", submodule.url.urlString, submodule.path],
+							repositoryFileURL: repositoryFileURL
+						)
+						// A .failure to add usually means the folder was already added
+						// to the index. That's okay.
+						.flatMapError { _ in SignalProducer<String, CarthageError>.empty }
+
+					// If it doesn't exist, clone and initialize a submodule from our
+					// local bare repository.
+					return cloneRepository(fetchURL, submoduleDirectoryURL, isBare: false)
+						.then(launchGitTask([ "remote", "set-url", "origin", submodule.url.urlString ], repositoryFileURL: submoduleDirectoryURL))
+						.then(checkoutRepositoryToDirectory(submoduleDirectoryURL, submoduleDirectoryURL, force: false))
+						.then(addSubmodule)
+						.then(launchGitTask([ "submodule", "--quiet", "init", "--", submodule.path ], repositoryFileURL: repositoryFileURL))
+						.startOnQueue(RepositoryOperationQueues.queue(forLocalRepositoryURL: submoduleDirectoryURL))
+						.then(self.updateSubmodules(submoduleDirectoryURL))
+				}
+			}
+	}
+
+	public func updateSubmodules(_ superprojectDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
+		func changeGitmodulesUrls(toCacheURL: Bool) -> SignalProducer<(), CarthageError> {
```

Comment:
I don't think this is necessary. The original code in `addSubmoduleToRepository` already handled doing the equivalent of `git submodule update --init` from the repository cache.

If your intent is to make that logic recursive, I think you can do that following the existing method without modifying `.gitmodules`.